virtwho/config.py

```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
+            result = ('warning', 'interval was not set to a valid integer: %s' % str(e))
+        return result
+
+    def _validate_str_to_bool(self, key):
+        result = None
+        try:
+            self._values[key] = str_to_bool(self._values[key])
+        except (KeyError, ValueError):
+            if self.has_default(key):
+                self._values[key] = str_to_bool(self.defaults[key])
+            result = ('warning', '%s must be a valid boolean, using default. '
+                                                   'See man virt-who-config for more info')
+        return result
+
+    def _validate_configs(self):
+        result = None
+        if self.is_default('configs'):
+            self._values['configs'] = []
+        elif isinstance(self['configs'], list):
+            self._values['configs'] = [item for item in self._values['configs'] if isinstance(item,
+                                                                                          str)]
+        elif not isinstance(self._values['configs'], str):
+            result = ('warning', '"configs" must be one or more strings, '
+                                                   'ignoring')
+            self._values['configs'] = []  # Reset to empty list
+        else:
+            self._values['configs'] = self._values['configs'].split(',')
+        return result
+
+    def _validate_non_empty_string(self, key):
+        result = None
+        if not isinstance(self._values[key], str):
+            result = ('warning', '%s is not set to a valid string, using default' % key)
+        elif len(self._values[key]) == 0:
+            result = ('warning', '%s cannot be empty, using default' % key)
+        return result
+
+    def validate(self, validation_messages=None):
+        validation_messages = super(GlobalSection, self).validate(validation_messages)
+        to_reset = set()
+        # Validate those keys that need to be validated
+        for key in set(self._unvalidated_keys):
+            error = None
+            # If boolean
+            if key in ['debug', 'oneshot', 'print', 'log_per_config', 'background']:
+                error = self._validate_str_to_bool(key)
+            elif key == 'configs':
+                error = self._validate_configs()
+            elif key in ['reporter_id', 'log_file', 'log_dir']:
+                error = self._validate_non_empty_string(key)
+            elif key == 'interval':
+                error = self._validate_interval()
+            else:
+                validation_messages.append(('warning', 'Ignoring unknown configuration option '
+                                                       '"%s"' % key))
+                del self._values[key]
+            if error is not None:
+                validation_messages.append(error)
+                if key not in ['interval']:  # Special cases not reset to default on failure
+                    to_reset.add(key)
+                self._invalid_keys.add(key)
```

Comment:
Yea I could see that I'll figure out another name. The intention behind it is a set of keys whose values are invalid.